src/job_metadata.c

```diff
@@ -793,47 +791,18 @@ cron_job_cache_invalidate(PG_FUNCTION_ARGS)
 static void
 InvalidateJobCache(void)
 {
-	HeapTuple classTuple = NULL;
-
-	classTuple = SearchSysCache1(RELOID, ObjectIdGetDatum(CronJobRelationId()));
-	if (HeapTupleIsValid(classTuple))
-	{
-		CacheInvalidateRelcacheByTuple(classTuple);
-		ReleaseSysCache(classTuple);
-	}
-}
-
-
-/*
- * InvalidateJobCacheCallback invalidates the job cache in response to
- * an invalidation event.
- */
-static void
-InvalidateJobCacheCallback(Datum argument, Oid relationId)
-{
-	if (relationId == CachedCronJobRelationId ||
-		CachedCronJobRelationId == InvalidOid)
-	{
-		CronJobCacheValid = false;
-		CachedCronJobRelationId = InvalidOid;
-	}
+	if (!pg_atomic_unlocked_test_flag_impl(CronJobCacheValid))
+		pg_atomic_clear_flag_impl(CronJobCacheValid);
```

Comment:
I couldn't quite follow this approach. Shared memory does not follow transaction boundaries (we will see the valid flag become false before commit), whereas invalidations do.

Reply:
> I couldn't quite follow this approach. Shared memory does not follow transaction boundaries (we will see the valid flag become false before commit), whereas invalidations do.

Yes, I have dived into the invalidation method, when the transaction commited, the pg_cron laucher can receive the invalidation messages and call 'InvalidateJobCacheCallback' to set the 'CronJobCacheValid' to be false. And when the transaction is aborted, pg_cron launcher cannot receive any messages. 
I agree that using the flag in shared memory can not follow transaction boundaries, although I can use a map to store the 'transactionID' and  the '  CronJobCacheValid' value  for each transaction and do something in at then end of the transaction,  it looks a little complex.

Reply:
@marcoslot I add a new commit in this pr, using xact_callback to catch table change event